File: Remove Nth Node from End of List.cpp
// Leetcode Problem 19: Remove Nth Node from End of List
// C++ CODE

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* first = dummy;
        ListNode* second = dummy;

        // Move first pointer n steps ahead
        for (int i = 0; i < n; i++) {
            first = first->next;
        }

        // Move both pointers until first reaches the end
        while (first->next != nullptr) {
            first = first->next;
            second = second->next;
        }

        // Remove the nth node from the end
        ListNode* nodeToRemove = second->next;
        second->next = second->next->next;
        delete nodeToRemove;

        return dummy->next;
    }
};

File: Remove Nth Node from End of List.py
# Leetcode Problem 19: Remove Nth Node From End of List
# Python CODE

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        first = dummy
        second = dummy

        # Move first pointer n steps ahead
        for _ in range(n):
            first = first.next

        # Move both pointers until first reaches the end
        while first.next:
            first = first.next
            second = second.next

        # Remove the nth node from the end
        second.next = second.next.next

        return dummy.next